### PR TITLE
Allow HPA to work with https heapster

### PR DESCRIPTION
**Pull request by liggitt**
Required for https://bugzilla.redhat.com/show_bug.cgi?id=1277914

### CODE REVIEW

**Comment by liggitt** [comment]
cc @deads2k @DirectXMan12

**Comment by liggitt** [comment]
I feel like we should expose options for heapster ns, scheme, service and port

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/services.go [hunk 1]
```diff
@@ -36,7 +37,7 @@ type ServiceInterface interface {
 	Update(srv *api.Service) (*api.Service, error)
 	Delete(name string) error
 	Watch(label labels.Selector, field fields.Selector, resourceVersion string) (watch.Interface, error)
-	ProxyGet(name, path string, params map[string]string) ResponseWrapper
+	ProxyGet(scheme, name, port, path string, params map[string]string) ResponseWrapper
```
Comment (deads2k) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/services.go]: These seem to be a commonly used tuple.  Why don't they have a type?

**Comment by deads2k** [comment]
> I feel like we should expose options for heapster ns, scheme, service and port

Not for 1.1.0.

**Comment by deads2k** [comment]
I dislike string tuples proliferating, but that's probably a larger fix.

lgtm

**Comment by liggitt** [comment]
>> I feel like we should expose options for heapster ns, scheme, service and port
>
> Not for 1.1.0.

I don't know the scenarios well enough to know that heapster will ONLY be served over https... it obviously wasn't for all the testing/dev work up to this point

**Comment by liggitt** [comment]
@smarterclayton @DirectXMan12 @mwringe feedback on requiring https for HPA->Heapster connections in origin?

**Comment by smarterclayton** [comment]
Is there a reason not to require it?

**Comment by mwringe** [comment]
Its been deemed a security issue to leave any unauthorized access to heapster open. This is why the origin-metrics heapster image requires ssl and certificate based authentication.

If we need to disable this security, its probably not that hard to add a configuration option. But we would really need make it clear what it means in the docs to disable it.

**Comment by smarterclayton** [comment]
I would be very uncomfortable leaving heapster unprotected unless it breaks the release.

**Comment by liggitt** [comment]
I'm fine with defaulting to https, but I don't know all the scenarios well enough to know if there are ever cases where we'd need to use http. All of upstream does, obviously, and all our dev and test to this point has been http

**Comment by smarterclayton** [comment]
does defaulting break existing API behavior?  If so we can't change the default, and I assumed this change is required?

**Comment by liggitt** [comment]
> does defaulting break existing API behavior? If so we can't change the default, and I assumed this change is required?

This isn't an API, it's how we set up the metrics client in the API server to talk to heapster. We can hard code `https` or make it configurable defaulting to `https`

**Comment by smarterclayton** [comment]
Ah, hardcode for 3.1 to https seems like it's ok for metrics client, and we
can fix up later.

On Nov 6, 2015, at 4:47 PM, Jordan Liggitt <notifications@github.com> wrote:

does defaulting break existing API behavior? If so we can't change the
default, and I assumed this change is required?

This isn't an API, it's how we set up the metrics client in the API server
to talk to heapster. We can hard code https or make it configurable
defaulting to https

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5763#issuecomment-154551469>.

**Comment by liggitt** [comment]
[test]

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/services.go (hunk 1 above)
Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/services.go]: Like... url.URL?

**Comment by smarterclayton** [comment]
In the upstream pull take a url.URL for proxy get, but I'll merge this as is.  It means we won't run in origin against http, but that's acceptable for 3.1 for me.

LGTM [merge]

**Comment by smarterclayton** [comment]
Let's get a follow up issue to make heapster configurable for 1.1.1

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/services.go (hunk 1 above)
Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/services.go]: where host is service:port?

Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/services.go]: If we're proxying a URL, I'd use a URL.  Host +port is a detail of config,
people using proxy get today are assuming port = scheme anyway.

On Nov 7, 2015, at 9:55 AM, Jordan Liggitt <notifications@github.com> wrote:

In
Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/services.go
<https://github.com/openshift/origin/pull/5763#discussion_r44213668>:

> @@ -36,7 +37,7 @@ type ServiceInterface interface {
>  	Update(srv *api.Service) (*api.Service, error)
>  	Delete(name string) error
>  	Watch(label labels.Selector, field fields.Selector, resourceVersion string) (watch.Interface, error)
> -	ProxyGet(name, path string, params map[string]string) ResponseWrapper
> +	ProxyGet(scheme, name, port, path string, params map[string]string) ResponseWrapper

where host is service:port?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5763/files#r44213668>.

**Pull request merged**